<a href="https://colab.research.google.com/github/Nico-GP/Coding-Dojo_Track2/blob/main/Week6/Abalone_Train_Test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **CODING DOJO**: KNN Abalon Train-Test
## Creado por: Nicolas Gonzalez
### Descripcion: Ejercicio de prediccion con KNN de edad y sexo de abalon sin y con division de entrenamiento
* Creado: 04/02/22
* Ultima Actualizacion: 04/02/22

In [82]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [83]:
import pandas as pd
ab = pd.read_csv('/content/drive/My Drive/CodingDojo/Week6/abalone.csv')
ab.head()

M  0.455  0.365  0.095   0.514  0.2245   0.101   0.15  15
0  M  0.350  0.265  0.090  0.2255  0.0995  0.0485  0.070   7
1  F  0.530  0.420  0.135  0.6770  0.2565  0.1415  0.210   9
2  M  0.440  0.365  0.125  0.5160  0.2155  0.1140  0.155  10
3  I  0.330  0.255  0.080  0.2050  0.0895  0.0395  0.055   7
4  I  0.425  0.300  0.095  0.3515  0.1410  0.0775  0.120   8

In [84]:
ab.shape

(4176, 9)

Se nombran las columnas de la base de datos

In [85]:
ab.columns =['Sex', 'Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight', 'Shell weight', 'Rings']
ab.head()

Sex  Length  Diameter  ...  Viscera weight  Shell weight  Rings
0   M   0.350     0.265  ...          0.0485         0.070      7
1   F   0.530     0.420  ...          0.1415         0.210      9
2   M   0.440     0.365  ...          0.1140         0.155     10
3   I   0.330     0.255  ...          0.0395         0.055      7
4   I   0.425     0.300  ...          0.0775         0.120      8

[5 rows x 9 columns]

In [86]:
ab.shape

(4176, 9)

In [87]:
import numpy as np

Se crea la columna de edad en años ("Age" =  "Rings" + 1.5)

In [88]:
ab.apply(lambda row: row.Rings + 1.5, axis=1)
ab['Age'] = ab.apply(lambda row: row.Rings + 1.5, axis=1)
ab.head()

Sex  Length  Diameter  Height  ...  Viscera weight  Shell weight  Rings   Age
0   M   0.350     0.265   0.090  ...          0.0485         0.070      7   8.5
1   F   0.530     0.420   0.135  ...          0.1415         0.210      9  10.5
2   M   0.440     0.365   0.125  ...          0.1140         0.155     10  11.5
3   I   0.330     0.255   0.080  ...          0.0395         0.055      7   8.5
4   I   0.425     0.300   0.095  ...          0.0775         0.120      8   9.5

[5 rows x 10 columns]

In [89]:
ab.isnull().sum()

Sex               0
Length            0
Diameter          0
Height            0
Whole weight      0
Shucked weight    0
Viscera weight    0
Shell weight      0
Rings             0
Age               0
dtype: int64

In [90]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [91]:
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_selection import RFE
from sklearn.svm import SVR
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

#Prediccion de edad

Se seleccionan las mediciones deseadas

In [92]:
ab_med = ab.loc[:, ['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight', 'Shell weight']]
ab_med.head()

Length  Diameter  Height  ...  Shucked weight  Viscera weight  Shell weight
0   0.350     0.265   0.090  ...          0.0995          0.0485         0.070
1   0.530     0.420   0.135  ...          0.2565          0.1415         0.210
2   0.440     0.365   0.125  ...          0.2155          0.1140         0.155
3   0.330     0.255   0.080  ...          0.0895          0.0395         0.055
4   0.425     0.300   0.095  ...          0.1410          0.0775         0.120

[5 rows x 7 columns]

In [93]:
medic = ['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight', 'Shell weight']
ab_med.loc[:, medic].head()

Length  Diameter  Height  ...  Shucked weight  Viscera weight  Shell weight
0   0.350     0.265   0.090  ...          0.0995          0.0485         0.070
1   0.530     0.420   0.135  ...          0.2565          0.1415         0.210
2   0.440     0.365   0.125  ...          0.2155          0.1140         0.155
3   0.330     0.255   0.080  ...          0.0895          0.0395         0.055
4   0.425     0.300   0.095  ...          0.1410          0.0775         0.120

[5 rows x 7 columns]

Se crea matriz de mediciones

In [94]:
X = ab_med.loc[:, medic].to_numpy()

In [95]:
X.shape

(4176, 7)

Se crea vector objetivo

In [96]:
y = ab.loc[:, 'Age'].values

In [97]:
y.shape

(4176,)

###Prediccion regresion lineal **sin** division train-test

In [98]:
reg = LinearRegression()
reg.fit(X, y)

LinearRegression()

Coeficiente de determinacion (R2)

In [99]:
reg.score(X, y)

0.5282432555640287

Raíz del error cuadrático medio (MSE)

In [100]:
np.sqrt(mean_squared_error(y, reg.predict(X)))

2.2138510563638767

###Prediccion regresion lineal **con** division train-test

In [101]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 3)

In [102]:
reg_split = LinearRegression()
reg_split.fit(X_train, y_train)

LinearRegression()

Coeficiente de determinacion (R2)

In [103]:
reg_split.score(X_train, y_train)

0.5283226055695078

In [104]:
reg_split.score(X_test, y_test)

0.518778205551305

Raíz del error cuadrático medio (MSE)

In [105]:
np.sqrt(mean_squared_error(y_train, reg_split.predict(X_train)))

2.2640990972285646

In [106]:
np.sqrt(mean_squared_error(y_test, reg_split.predict(X_test)))

2.061108906779863

###Prediccion KNN **sin** division train-test

Se llama a la regresion K

In [107]:
scaler.fit(X)
X_scal = scaler.transform(X)

In [118]:
knn = KNeighborsRegressor()
knn.fit(X_scal, y)

KNeighborsRegressor()

Coeficiente de determinacion (R2)

In [119]:
knn.score(X_scal, y)

0.6582492859136977

Raíz del error cuadrático medio (MSE)

In [120]:
np.sqrt(mean_squared_error(y, knn.predict(X_scal)))

1.8842739996473779

###Prediccion KNN **con** division train-test

In [121]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 3)

In [122]:
scaler.fit(X_train)
X_scal = scaler.transform(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

In [123]:
knn_split = KNeighborsRegressor()
knn_split.fit(X_train, y_train)

KNeighborsRegressor()

Coeficiente de determinacion (R2)

In [125]:
knn_split.score(X_train, y_train)

0.6555512962081227

In [126]:
knn_split.score(X_test, y_test)

0.4714544304886764

Raíz del error cuadrático medio (MSE)

In [127]:
np.sqrt(mean_squared_error(y_train, knn_split.predict(X_train)))

1.9347959852320313

In [128]:
np.sqrt(mean_squared_error(y_test, knn_split.predict(X_test)))

2.1600784007341045

#Preguntas

¿Cuál de KNN o regresión lineal pareció un mejor modelo cuando no usaste la división entrenar probar?

* KNN, ya que tuvo un indice de acierto mayor (R2 - 65% vs 52%) y un error absoluto menor (MSE - 1.8 vs 2.2)

¿Cuál de KNN o regresión lineal pareció un mejor modelo cuando usaste la división entrenar probar?

* Regresion lineal, ya que tuvo un indice de acierto mayor en la etapa de testeo (R2 - 51% vs 47%) y un error absoluto menor (MSE - 2.0 vs 2.1)

¿Hubo alguna ventaja en la regresión lineal en términos de la cantidad de código que tenías que escribir?

* Regresion lineal tiene un codigo ligeramente menos extenso que KNN ya que no hay que escalar

¿Hay alguna forma de mostrarle a alguien cuál de los dos modelos fue más efectivo?

* Comparando las métricas (R2, MAE, MSE, RMSE), especialmente las de testeo

¿Crees que podrías haber mejorado KNN para mejorar la eficacia del modelo?

* Se pueden modificar los parametros. Por ejemplo el numero de vecinos